src/libs/ease/hash.c

```diff
@@ -0,0 +1,118 @@
+/**
+ * @file
+ *
+ * @brief Provides functions to hash Elektra data structures.
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ */
+
+#include <kdb.h>
+#include <kdbease.h>
+#include <kdbtypes.h>
+#include <kdberrors.h>
+
+#include <string.h>
+#include <stdio.h>
+
+#include "sha-256.h"
+
+static void hash_to_string(char string[65], const uint8_t hash[32]);
+
+/**
+ * Calculate a specification token for the KeySet of an application.
+ *
+ * The KeySet of an application is identified as all keys below the applications root key.
+ *
+ * @pre The parentKey must have the correct namespace. E.g. If only keys from the spec:/ should be considered for the token calculation,
+ * pass a key with KEY_NS_SPEC.
+ *
+ * @param hash_string A string. After successful execution this will contain the hash as hex-string.
+ * @param ks The KeySet for the application.
+ * @param parentKey The Key below which all the relevant keys are. Keys that are not below @p parentKey are ignored. The key's namespace is important (see preconditions)
+ * @retval false If an error occurred.
+ * @retval true If the computation was successful.
+ */
+kdb_boolean_t calculateSpecificationToken (char hash_string[65], KeySet * ks, Key * parentKey) {
+	if(parentKey == NULL) {
+		// Can't set error to parentKey when it is null.
+		return false;
+	}
+	if(hash_string == NULL) {
+		ELEKTRA_SET_INTERNAL_ERROR(parentKey, "Param hash_string was NULL");
+		return false;
+	}
+	if(ks == NULL) {
+		ELEKTRA_SET_INTERNAL_ERROR(parentKey, "Param ks was NULL");
+		return false;
+	}
+
+	// Initialize sha_256 for streaming
+	uint8_t hash[SIZE_OF_SHA_256_HASH];
+	struct Sha_256 sha_256;
+	sha_256_init(&sha_256, hash);
+
+	// Duplicate ks, then cut out parentKey and all keys below. These are the ones we take into account for token calculation.
+	KeySet * dupKs = ksDup (ks);
+	KeySet * cutKs = ksCut (dupKs, parentKey);
+
+	/**
+	 * Loop through all keys relevant for token calculation.
+ 	*/
+	Key * currentKey;
+	ksRewind(cutKs);
+	while ((currentKey = ksNext (cutKs)) != NULL) {
+		Key * cascadingKey = keyDup (currentKey, KEY_CP_NAME);
+		if(cascadingKey == NULL) {
+			ELEKTRA_SET_INTERNAL_ERROR (parentKey, "keyDup() failed!");
+			return false;
+		}
+		/**
+		 * Key namespace is ignored for token calculation (via setting it to cascading).
+		 * Reason: Different callers of this function pass the keys for token calculation using different namespaces,
+		 * but the specification they pass is actually the same.
+		 * (e.g. tools/kdb/gen.cpp passes keys in cascading namespace while src/libs/highlevel/elektra.c passes keys in spec namespace).
+		 */
+		int result = keySetNamespace (cascadingKey, KEY_NS_CASCADING);
```

Comment:
Is this already the new algorithm?

Reply:
Yes, this is the algorithm we discussed yesterday!

Reply:
Yes, that is correct!

* We only look at spec: keys in the algorithm
* But, when we feed the keyname into the sha-256 API, we want to ignore the namespace prefix. The reason is: When `kdb gen` is used with a specification file as input, these keys are all cascading (even though they are spec: keys). If we don't remove the prefix everytime, the token will not match for `kdb gen` (cascading keys) vs. application startup (spec: keys).

Reply:
I don't understand why the namespace gets removed. I think we should pass everything as unmodified as possible to the sha calculator.

Reply:
I think, if my rationale can't be understood from the code + inline comments + further explanations here, I'll have to change the code :)

I'll try to find another solution for the namespace problem!

Reply:
I don't think another solution is needed! I only wonder which other caller has a different namespace? I thought that we only compare specs? Is the idea that this function should be able to be used more general than we currently need? Such information (like that KeySets from any namespace can be compared)) should be in the comments for the function, not somewhere hidden inline.

Reply:
I'm also not sure, you need to ignore the namespace. Even if you want to ignore the namespace, you don't need to `keyDup` the key for that. First you can use the unescaped name instead of the escaped one. Then you could simply do:

```c
// ignore the first byte, which is always the namespace
sha_256_write (&sha_256, keyUnescapedName (currentKey) + 1, keyGetUnescapedNameSize (currentKey) - 2);
```

Using the unescaped name makes it a bit harder to compare against externally calculated hashes. But you could do something similar with `strchr` for the escaped name, by looking for the first `/`. This makes the length calculation a bit more difficult, but it avoid duplicating every key.

Reply:
Thank you both, I fixed this and the namespace is now not ignored anymore.